util/buildinfo/buildinfo.go

```diff
@@ -2,136 +2,208 @@ package buildinfo
+	// don't need but there is an edge case if no instruction except sources
+	// one is defined (e.g. FROM ...) that can be valid so take it into account.
+	for _, ics := range imageConfigBuildInfo.Sources {
+		if ics.Type != binfotypes.SourceTypeDockerImage {
 			continue
 		}
-		if _, ok := mbis[bi.Alias]; !ok {
-			parsed, err := reference.ParseNormalizedNamed(bi.Ref)
+		if _, ok := mbs[ics.Alias]; !ok {
+			parsed, err := reference.ParseNormalizedNamed(ics.Ref)
 			if err != nil {
-				return nil, errors.Wrapf(err, "failed to parse %s", bi.Ref)
+				return nil, errors.Wrapf(err, "failed to parse %s", ics.Ref)
 			}
-			mbis[bi.Alias] = exptypes.BuildInfo{
-				Type: exptypes.BuildInfoTypeDockerImage,
+			mbs[ics.Alias] = binfotypes.Source{
+				Type: binfotypes.SourceTypeDockerImage,
 				Ref:  reference.TagNameOnly(parsed).String(),
-				Pin:  bi.Pin,
+				Pin:  ics.Pin,
 			}
 		}
 	}
 
-	bis := make([]exptypes.BuildInfo, 0, len(mbis))
-	for _, bi := range mbis {
-		bis = append(bis, bi)
+	srcs := make([]binfotypes.Source, 0, len(mbs))
+	for _, bs := range mbs {
+		srcs = append(srcs, bs)
 	}
-	sort.Slice(bis, func(i, j int) bool {
-		return bis[i].Ref < bis[j].Ref
+	sort.Slice(srcs, func(i, j int) bool {
+		return srcs[i].Ref < srcs[j].Ref
 	})
 
-	return json.Marshal(map[string][]exptypes.BuildInfo{
-		"sources": bis,
-	})
+	return srcs, nil
 }
 
-// imageConfigBuildInfo returns build dependencies from image config
-func imageConfigBuildInfo(imageConfig []byte) ([]exptypes.BuildInfo, error) {
+// FromImageConfig returns build dependencies from image config.
+func FromImageConfig(imageConfig []byte) (bi binfotypes.BuildInfo, _ error) {
 	if len(imageConfig) == 0 {
-		return nil, nil
-	}
-	var config struct {
-		BuildInfo []byte `json:"moby.buildkit.buildinfo.v1,omitempty"`
+		return bi, nil
 	}
+	var config binfotypes.ImageConfig
 	if err := json.Unmarshal(imageConfig, &config); err != nil {
-		return nil, errors.Wrap(err, "failed to unmarshal buildinfo from config")
+		return bi, errors.Wrap(err, "failed to unmarshal buildinfo from image config")
 	}
 	if len(config.BuildInfo) == 0 {
-		return nil, nil
+		return bi, nil
 	}
-	var bi []exptypes.BuildInfo
 	if err := json.Unmarshal(config.BuildInfo, &bi); err != nil {
-		return nil, errors.Wrapf(err, "failed to unmarshal %s", ImageConfigField)
+		return bi, errors.Wrap(err, "failed to unmarshal buildinfo")
 	}
 	return bi, nil
 }
+
+var knownAttrs = []string{
+	"cmdline",
+	"context",
+	"filename",
+	"source",
+
+	//"add-hosts",
+	//"cgroup-parent",
+	//"force-network-mode",
+	//"hostname",
+	//"image-resolve-mode",
+	"platform",
+	"shm-size",
+	"target",
+	"ulimit",
+}
+
+// filterAttrs filters frontent opt by picking only those that
+// could effectively change the build result.
+func filterAttrs(attrs map[string]string) map[string]string {
+	filtered := make(map[string]string)
+	for k, v := range attrs {
+		if strings.HasPrefix(k, "build-arg:") || strings.HasPrefix(k, "label:") {
```

Comment:
For the `BUILDKIT_` prefixed control args, maybe better to have a known list also?

Reply:
Yes indeed with should handle them.